ssl/ssl_key_share.cc

```diff
@@ -176,20 +180,557 @@ class X25519KeyShare : public SSLKeyShare {
+// all of which are used to generate a hybrid shared secret.
+// See https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design.
+  class HybridKeyShare : public SSLKeyShare {
+  public:
+    HybridKeyShare(uint16_t group_id) : group_id_(group_id),
+    exchange_performed(false), hybrid_group_(nullptr) {
+      for (const HybridGroup &hybrid_group : HybridGroups()) {
+         if (group_id_ == hybrid_group.group_id) {
+           hybrid_group_ = &hybrid_group;
+           for (size_t i = 0; i < NUM_HYBRID_COMPONENTS; i++) {
+             key_shares_[i] = SSLKeyShare::Create(hybrid_group.component_group_ids[i]);
+           }
+           return;
+         }
+      }
+      hybrid_group_ = nullptr;
+    }
+
+    uint16_t GroupID() const override { return group_id_; }
+
+    bool Offer(CBB *out) override {
+      // Ensure |out| is valid, has no children, and has been initialized.
+      // We check that |out| has no children because otherwise CBB_data()
+      // will produce a fatal error by way of assert(cbb->child == NULL);
+      if (!out || out->child || !CBB_data(out)) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      if (!hybrid_group_ || this->exchange_performed) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Iterate through the component groups and Offer() each of their key
+      // shares. If one of the calls to a component Offer() fails,
+      // OPENSSL_PUT_ERROR will be set appropriately in that component.
+      for (const UniquePtr<SSLKeyShare> &key_share : key_shares_) {
+        if (!key_share || !key_share->Offer(out)) {
+          return false;
+        }
+      }
+      this->exchange_performed = true;
+      return true;
+    }
+
+    bool Accept(CBB *out_public_key, Array<uint8_t> *out_secret,
+                uint8_t *out_alert, Span<const uint8_t> peer_key) override {
+      if (!out_alert) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      // Set alert to internal error by default
+      *out_alert = SSL_AD_INTERNAL_ERROR;
+
+      if (!out_secret|| !peer_key.data()) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      // Ensure |out_public_key| is valid, has no children, and has been
+      // initialized. We check that |out_public_key| has no children because
+      // otherwise CBB_data() will produce a fatal error by way of
+      // assert(cbb->child == NULL);
+      if (!out_public_key || out_public_key->child || !CBB_data(out_public_key)) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      if (!hybrid_group_ || this->exchange_performed) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // A hybrid shared secret with two components should be 64 bytes.
```

Comment:
where does 64 come from? is it defined in some spec?

Reply:
It is the initial capacity of the Crypto Byte Buffer that will be storing the hybrid secret in. It will grow as needed as data is added.